### PR TITLE
Move registration code to a distinct service.

### PR DESCRIPTION
**Pull request by awood**

### CODE REVIEW

File: src/rhsmlib/services/register.py [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+    def __init__(self, cp):
+        self.plugin_manager = inj.require(inj.PLUGIN_MANAGER)
+        self.installed_mgr = inj.require(inj.INSTALLED_PRODUCTS_MANAGER)
+        self.facts = inj.require(inj.FACTS)
+        self.identity = inj.require(inj.IDENTITY)
+        self.cp = cp
+
+    def register(self, org, activation_keys=None, environment=None, force=None, name=None, consumerid=None,
+            **kwargs):
+        # Strip out args set to None since validate_options looks for presence in the dictionary
```
Comment (kahowell) [inline src/rhsmlib/services/register.py]: Looks like this comment goes with line 48?

**Comment by kahowell** [comment]
Noticed some weird behavior when testing this. Confusing behavior when already registered and not recognizing already registered on a second call to register... Will dig a little deeper and leave better feedback tomorrow.

File: src/rhsmlib/services/register.py [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+            # 746259: Don't allow the user to pass in an empty string as an activation key
+            if '' == options['activation_keys']:
+                error_msg = _("Error: Must specify an activation key")
+            elif getattr(self.cp, 'username', None) or getattr(self.cp, 'password', None):
+                error_msg = _("Error: Activation keys do not require user credentials.")
+            elif options['consumerid']:
+                error_msg = _("Error: Activation keys can not be used with previously registered IDs.")
+            elif options['environment']:
+                error_msg = _("Error: Activation keys do not allow environments to be specified.")
+        elif not getattr(self.cp, 'username', None) or not getattr(self.cp, 'password', None):
```
Comment (kahowell) [inline src/rhsmlib/services/register.py]: In the case that you're already registered, this branch overrides the more accurate error message already set ("This system is already registered. Add force to options to override."), as `build_uep` gives a connection using the identity cert if it exists, regardless of whether username/password is passed.

File: src/rhsmlib/services/register.py [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+        else:
+            consumer = self.cp.registerConsumer(
+                name=consumer_name,
+                facts=facts_dict,
+                owner=org,
+                environment=environment,
+                keys=options.get('activation_keys'),
+                installed_products=self.installed_mgr.format_for_server(),
+                content_tags=self.installed_mgr.tags
+            )
```
Comment (kahowell) [inline src/rhsmlib/services/register.py]: The identity cert (`self.identity` in this context) needs a call to `reload` at some point. Right after this statement or in `managerlib.persist_consumer_cert` might be appropriate. Without the reload, if I call register twice it doesn't acknowledge that I'm already registered, and lets me repeatedly call register.

Comment (awood) [inline src/rhsmlib/services/register.py]: Addressed this but just not on this specific line.

**Comment by awood** [comment]
> Re-tested, and I think the order of operations is still slightly off. After registering twice, on the third attempt, I get the "force is necessary" message. persist_consumer_cert needs to happen before the reload.

Good catch.  I've moved the certificate persistence into the service itself since that's where it really belongs.

**Pull request merged**